pkg/operator/baremetal_config.go

```diff
@@ -0,0 +1,63 @@
+package operator
+
+import (
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	dynamic "k8s.io/client-go/dynamic"
+)
+
+var provisioningGVR = schema.GroupVersionResource{Group: "metal3.io", Resource: "provisionings", Version: "v1alpha1"}
+var configName = "test"
+
+func getBaremetalProvisioningConfig(dc dynamic.Interface, configName string, usingBareMetal bool) BaremetalConfig {
+	if !usingBareMetal {
+		return BaremetalConfig{}
+	}
+	provisioningClient := dc.Resource(provisioningGVR)
+	provisioningConfig, err := provisioningClient.Get(configName, metav1.GetOptions{})
+	if err != nil {
+		glog.V(1).Infof("Error getting provisioning config %q", configName)
+		return BaremetalConfig{}
+	}
+	provisioningSpec, found, err := unstructured.NestedMap(provisioningConfig.UnstructuredContent(), "spec")
+	if !found {
+		glog.Errorf("Nested Spec not found in Baremetal provisioning CR %q", configName)
```

Comment:
don't you want to error here for all the bad cases? I'd suggest you move this function to be called by https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/sync.go#L77 instead of the current caller you can handle errors as you wish without impacting other platforms.
Also how is this going to handle defaults?

Reply:
The current implementation will have to support reading from the CR first and if that fails, fall back to reading from the ConfigMap to support backwards compatibility among other things. So, when we hit errors, we plan to just return an empty BaremetalConfig struct. Logic that is coming in a following PR would see that we were not able to get config from the CR and fallback to  using the configmap.
https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/baremetal_pod.go#L17.

Regarding calling getBaremetalProvisioningConfig() from sync.go (https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/sync.go#L77 ). We added BaremetalConfig to OperatorConfig and ideally before we try to deploy metal3, OperatorConfig should be fully populated.
We could potentially, leave BaremetalConfig uninitialized until we try to deploy metal3. I was keeping the flow similar to the other baremetal related config, the BaremetalControllers.

Reply:
I see, so if you fail to read the CRD, or there's a missing field in the CRD, then you fallback to the configMap and if anything fails you return the empty struct. What if there's a transient failure to read from the API you just set the empty struct as well?
In any case I'd still prefer this to be in https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/sync.go#L77 so this is completely isolated from other platforms and you can change how to handle those errors anytime.

Reply:
That is correct. Updating to accommodate comment.

Reply:
Refactored to remove BaremetalConfig from OperatorConfig and moved initialization of BaremetalConfig to within https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/sync.go#L77 .